src/errno.cr

```diff
@@ -16,232 +16,52 @@ end
-  EISCONN = LibC::EISCONN
-  # Socket is not connected
-  ENOTCONN = LibC::ENOTCONN
-  # Can't send after socket shutdown
-  ESHUTDOWN = LibC::ESHUTDOWN
-  # Too many references: can't splice
-  ETOOMANYREFS = LibC::ETOOMANYREFS
-  # Operation timed out
-  ETIMEDOUT = LibC::ETIMEDOUT
-  # Connection refused
-  ECONNREFUSED = LibC::ECONNREFUSED
-  # Too many levels of symbolic links
-  ELOOP = LibC::ELOOP
-  # File name too long
-  ENAMETOOLONG = LibC::ENAMETOOLONG
-  # Host is down
-  EHOSTDOWN = LibC::EHOSTDOWN
-  # No route to host
-  EHOSTUNREACH = LibC::EHOSTUNREACH
-  # Directory not empty
-  ENOTEMPTY = LibC::ENOTEMPTY
-  # Too many users
-  EUSERS = LibC::EUSERS
-  # Disc quota exceeded
-  EDQUOT = LibC::EDQUOT
-  # Stale NFS file handle
-  ESTALE = LibC::ESTALE
-  # Too many levels of remote in path
-  EREMOTE = LibC::EREMOTE
-  # No locks available
-  ENOLCK = LibC::ENOLCK
-  # Function not implemented
-  ENOSYS = LibC::ENOSYS
-  # Value too large to be stored in data type
-  EOVERFLOW = LibC::EOVERFLOW
-  # Operation canceled
-  ECANCELED = LibC::ECANCELED
-  # Identifier removed
-  EIDRM = LibC::EIDRM
-  # No message of desired type
-  ENOMSG = LibC::ENOMSG
-  # Illegal byte sequence
-  EILSEQ = LibC::EILSEQ
-  # Bad message
-  EBADMSG = LibC::EBADMSG
-  # Reserved
-  EMULTIHOP = LibC::EMULTIHOP
-  # No message available on STREAM
-  ENODATA = LibC::ENODATA
-  # Reserved
-  ENOLINK = LibC::ENOLINK
-  # No STREAM resources
-  ENOSR = LibC::ENOSR
-  # Not a STREAM
-  ENOSTR = LibC::ENOSTR
-  # Protocol error
-  EPROTO = LibC::EPROTO
-  # STREAM ioctl timeout
-  ETIME = LibC::ETIME
-  # Operation not supported on socket
-  EOPNOTSUPP = LibC::EOPNOTSUPP
-  # State not recoverable
-  ENOTRECOVERABLE = LibC::ENOTRECOVERABLE
-  # Previous owner died
-  EOWNERDEAD = LibC::EOWNERDEAD
+enum Errno
+  NONE = 0
 
-  # Returns the numeric value of errno.
-  getter errno : Int32
-
-  # Returns the message of errno.
-  getter errno_message : String
+  {% for value in %w(E2BIG EPERM ENOENT ESRCH EINTR EIO ENXIO ENOEXEC EBADF ECHILD EDEADLK ENOMEM
+                    EACCES EFAULT ENOTBLK EBUSY EEXIST EXDEV ENODEV ENOTDIR EISDIR EINVAL ENFILE
+                    EMFILE ENOTTY ETXTBSY EFBIG ENOSPC ESPIPE EROFS EMLINK EPIPE EDOM ERANGE EAGAIN
+                    EWOULDBLOCK EINPROGRESS EALREADY ENOTSOCK EDESTADDRREQ EMSGSIZE EPROTOTYPE ENOPROTOOPT
+                    EPROTONOSUPPORT ESOCKTNOSUPPORT EPFNOSUPPORT EAFNOSUPPORT EADDRINUSE EADDRNOTAVAIL
+                    ENETDOWN ENETUNREACH ENETRESET ECONNABORTED ECONNRESET ENOBUFS EISCONN ENOTCONN
+                    ESHUTDOWN ETOOMANYREFS ETIMEDOUT ECONNREFUSED ELOOP ENAMETOOLONG EHOSTDOWN
+                    EHOSTUNREACH ENOTEMPTY EUSERS EDQUOT ESTALE EREMOTE ENOLCK ENOSYS EOVERFLOW
+                    ECANCELED EIDRM ENOMSG EILSEQ EBADMSG EMULTIHOP ENODATA ENOLINK ENOSR ENOSTR
+                    EPROTO ETIME EOPNOTSUPP ENOTRECOVERABLE EOWNERDEAD) %}
+    {% if LibC.has_constant?(value) %}
```

Comment:
We should either move this enum into `Crystal::` so it is not public or have it be always the same list of errno despite the platform: this enum must have the same *members* regardless of platform.

Reply:
But not all errno values exist in all platforms. Maybe we can map those to EINVAL. But there should be at least a warning if you try to use any of those values in a platform where the value doesn't exist.

Reply:
We decided a long time ago that the compile-time API must be identical between platforms so that everything compiles on every platform.

If the errno value doesn't exist on every platform, it just never gets used, never gets raised, so it can be any value.

I don't think we need warnings for this.

Reply:
The existence of an Error number is one part of the problem. There is platform-specific value for some error codes. For example, `grep -R "EDEADLK" src/lib_c/*`

```
src/lib_c/aarch64-linux-gnu/c/errno.cr:        EDEADLK = 35
src/lib_c/aarch64-linux-musl/c/errno.cr:       EDEADLK = 35
src/lib_c/amd64-unknown-openbsd/c/errno.cr:    EDEADLK = 11
src/lib_c/arm-linux-gnueabihf/c/errno.cr:      EDEADLK = 35
src/lib_c/i386-linux-gnu/c/errno.cr:           EDEADLK = 35
src/lib_c/i386-linux-musl/c/errno.cr:          EDEADLK = 35
src/lib_c/i686-linux-gnu/c/errno.cr:           EDEADLK = 35
src/lib_c/i686-linux-musl/c/errno.cr:          EDEADLK = 35
src/lib_c/x86_64-darwin/c/errno.cr:            EDEADLK = 11
src/lib_c/x86_64-dragonfly/c/errno.cr:         EDEADLK = 11
src/lib_c/x86_64-freebsd/c/errno.cr:           EDEADLK = 11
src/lib_c/x86_64-linux-gnu/c/errno.cr:         EDEADLK = 35
src/lib_c/x86_64-linux-musl/c/errno.cr:        EDEADLK = 35
src/lib_c/x86_64-macosx-darwin/c/errno.cr:     EDEADLK = 11
src/lib_c/x86_64-openbsd/c/errno.cr:           EDEADLK = 11
src/lib_c/x86_64-portbld-freebsd/c/errno.cr:   EDEADLK = 11
src/lib_c/x86_64-unknown-freebsd/c/errno.cr:   EDEADLK = 11
src/lib_c/x86_64-windows-msvc/c/errno.cr:      EDEADLK = 36
```